### ISSUE TITLE
Error on system with 3.1 and 5

### ISSUE THREAD

**Issue by lonix1** [tintoy/msbuild-project-tools-server#21]
I'm trying to use this on a system with .NET5, but I saw that this isn't supported. So I installed the 3.1 runtime.

Now the extension loads, but I get an error:

> Error loading MSBuild project '"/foo/bar/myproject/src/Directory.Build.props"'.
> System.IO.FileNotFoundException: Could not load file or assembly 'System.Collections.Immutable, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

There is no problem with my config, as it works when I use `dotnet build`, `dotnet run`, etc.

What could this be?

**Comment by tintoy** [comment]
Hi, that does seem odd; we are using the new MSBuildLocator API so it should be loading the v5-compatible version of MSBuild where possible (if it’s reachable via a discovered version of the .NET Core SDK).

If you increase the logging level to verbose, is there a stack trace?

Does your .props include any custom tasks or function usages?

Also, can you include the output of `dotnet —info`?

**Comment by lonix1** [comment]
Verbose logging enabled. Log from "MSBuild Project Tools" output window:

````log
Result was "System.Threading.Tasks.Task`1[[OmniSharp.Extensions.LanguageServer.Protocol.Models.InitializeResult, OmniSharp.Extensions.LanguageProtocol, Version=0.7.9.0, Culture=neutral, PublicKeyToken=null]]"
Response value was "OmniSharp.Extensions.LanguageServer.Protocol.Models.InitializeResult"
Finished: Routing Request (0) "initialize" in 69ms
Finding descriptor for "initialized"
Starting: Routing Notification "initialized"
Finding descriptor for "workspace/didChangeConfiguration"
Finished: Routing Notification "initialized" in 9ms
Starting: Routing Notification "workspace/didChangeConfiguration"
Converting params for Notification "workspace/didChangeConfiguration" to "MSBuildProjectTools.LanguageServer.CustomProtocol.DidChangeConfigurationObjectParams"
Finished: Routing Notification "workspace/didChangeConfiguration" in 70ms
Language server initialised by client.


Finding descriptor for "textDocument/didOpen"
Created attribute "xml::file:///home/username/foo/project/Bar.csproj"
Looking for handler for method ["OmniSharp.Extensions.LanguageServer.Server.HandlerDescriptor"]
Checking handler ["OmniSharp.Extensions.LanguageServer.Server.HandlerDescriptor"]:"MSBuildProjectTools.LanguageServer.Handlers.DocumentSyncHandler"
Registration options "OmniSharp.Extensions.LanguageServer.Protocol.Models.TextDocumentRegistrationOptions"
Document Selector "[msbuild, file, **/*.*], [xml, file, **/*.*proj], [xml, file, **/*.props], [xml, file, **/*.targets]"
Starting: Routing Notification "textDocument/didOpen"
Converting params for Notification "textDocument/didOpen" to "OmniSharp.Extensions.LanguageServer.Protocol.Models.DidOpenTextDocumentParams"
[Info  - 9:45:45 AM] 1 package sources configured for project "/home/username/foo/project/Bar.csproj".
[Info  - 9:45:45 AM]   Locally-configured package source "nuget.org" (v3) => https://api.nuget.org/v3/index.json
Launching "dotnet --info"...
Launched "dotnet --info". Waiting for process 39512 to terminate...
"dotnet --info" terminated with exit code 0.
"dotnet --info" returned the following text on STDOUT / STDERR.

.NET SDK (reflecting any global.json):
 Version:   5.0.102
 Commit:    71365b4d42

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         ubuntu.20.04-x64
 Base Path:   /usr/share/dotnet/sdk/5.0.102/

Host (useful for support):
  Version: 5.0.2
  Commit:  cb5f173b96

.NET SDKs installed:
  5.0.102 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.11 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.11 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download



[Error - 9:45:46 AM] Error loading MSBuild project '"/home/username/foo/project/Bar.csproj"'.
System.IO.FileNotFoundException: Could not load file or assembly 'System.Collections.Immutable, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

File name: 'System.Collections.Immutable, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at new Microsoft.Build.Evaluation.Context.EvaluationContext(SharingPolicy policy, IFileSystem fileSystem)
   at EvaluationContext Microsoft.Build.Evaluation.Context.EvaluationContext.Create(SharingPolicy policy, MSBuildFileSystemBase fileSystem)
   at void Microsoft.Build.Evaluation.Project+ProjectImpl.Reevaluate(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at void Microsoft.Build.Evaluation.Project+ProjectImpl.ReevaluateIfNecessary(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at void Microsoft.Build.Evaluation.Project+ProjectImpl.ReevaluateIfNecessary(ILoggingService loggingServiceForEvaluation, EvaluationContext evaluationContext)
   at void Microsoft.Build.Evaluation.Project+ProjectImpl.ReevaluateIfNecessary(EvaluationContext evaluationContext)
   at void Microsoft.Build.Evaluation.Project+ProjectImpl.Initialize(IDictionary<string, string> globalProperties, string toolsVersion, string subToolsetVersion, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at new Microsoft.Build.Evaluation.Project(string projectFile, IDictionary<string, string> globalProperties, string toolsVersion, string subToolsetVersion, ProjectCollection projectCollection, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at Project Microsoft.Build.Evaluation.ProjectCollection.LoadProject(string fileName, IDictionary<string, string> globalProperties, string toolsVersion)
   at Project Microsoft.Build.Evaluation.ProjectCollection.LoadProject(string fileName, string toolsVersion)
   at Project Microsoft.Build.Evaluation.ProjectCollection.LoadProject(string fileName)
   at bool MSBuildProjectTools.LanguageServer.Documents.MasterProjectDocument.TryLoadMSBuildProject() in /home/username/foo/project/C:\Development\github\tintoy\msbuild-project-tools-vscode\lib\server\src\LanguageServer.Engine\Documents\MasterProjectDocument.cs:line 173


Not scanning package references (although existing references have been cleared) for MSBuild project "/home/username/foo/project/Bar.csproj" because the project is not currently loaded.
Requesting suggestions for NuGet package Ids matching prefix "Newtonsoft.Json" (include pre-release: False)...
===========================
 - Project uses package source "nuget.org" ("https://api.nuget.org/v3/index.json")
===========================
MSBuild project not loaded.
Finished: Routing Notification "textDocument/didOpen" in 570ms
Found 7 suggestions for NuGet package Ids matching prefix "Newtonsoft.Json" (include pre-release: False).
````

From `dotnet --info`:
````log
.NET SDK (reflecting any global.json):
 Version:   5.0.102
 Commit:    71365b4d42

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         ubuntu.20.04-x64
 Base Path:   /usr/share/dotnet/sdk/5.0.102/

Host (useful for support):
  Version: 5.0.2
  Commit:  cb5f173b96

.NET SDKs installed:
  5.0.102 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.11 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.11 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
````

**Comment by tintoy** [comment]
That’s strange - I’ll see if I can try this out again tomorrow to figure out what the problem is; it should work, as far as I know.

**Comment by lonix1** [comment]
Don't trouble yourself... it might just be me. If no one else has this problem, that's probably the reason! :)

**Comment by tintoy** [comment]
The MSBuild and SDK version treadmill is a constant source of issues, sadly - I’ll have a go at trying out the upgrade to 5.x tomorrow; if I can get you a pre-release .vsix package for the extension would you be able to try it out for me?

**Comment by lonix1** [comment]
I think vsix is for VS?, but I'm using VSCode...

Sure I'll try it, but I'll only be back at the office on Monday.

**Comment by tintoy** [comment]
Yep, VSCode extensions also use the vsix package format :)

Thanks - will see what I can do tomorrow and then let you know

**Comment by tintoy** [comment]
Ok, can you give this one a try and let me know if it works for you?

[msbuild-project-tools-0.3.12.vsix.zip](https://github.com/tintoy/msbuild-project-tools-server/files/5896764/msbuild-project-tools-0.3.12.vsix.zip)

**Comment by lonix1** [comment]
It works!

The log output shows:
````log
Starting MSBuild language service...
MSBuild language service is running.
````

No errors!

Hovering over tags gives me info. I get intellisense for package versions.

What else would you like me to test?

**Comment by tintoy** [comment]
Hi - sorry, the last couple of days have been a little hectic. I’ll do some testing with the 3.x SDK to make sure nothing has been broken there when using this new version and if it works then I’ll publish the new extension.

**Comment by tintoy** [comment]
Published v0.3.12 of the extension (which now targets 5.0).

Sorry for the delay; had to find time to set up VMs for testing with multiple versions :)

**Comment by lonix1** [comment]
Don't apologise for giving us this great free sofware! :smile: 

Works nicely on net5, thanks!
